### DISCUSSION TITLE
Avalonia insists on using App's default constructor on iOS

### DISCUSSION THREAD

**Discussion by AArnott** [AvaloniaUI/Avalonia#16382] [Q&A] [answered]
I have an App class with a default constructor (for design-time support) and a non-default constructor (for the actual runtime apps). This works great on Windows, Linux and MacOS, as well as Android. 

But on iOS, I use the same technique but it is ignored. Specifically, my `AppDelegate` class has this member:

```cs
protected override AppBuilder CustomizeAppBuilder(AppBuilder builder)
{
	// Discard the builder we're given because we want to call the non-default App constructor.
	builder = base.CustomizeAppBuilder(AppBuilder.Configure(
		() => new App(PrepareAppPlatformSettings(), new IosPlatformServices(this), null, null)))
		.WithInterFont()
		.UseReactiveUI()
		.UseiOS();

	return builder;
}
```

And while this method executes, the delegate I pass to `AppBuilder.Configure` never does. Instead, my App class's default constructor runs due to a call from `AppBuilder` to `Activator.CreateInstance`. 

I can't figure out why `AppBuilder` is ignoring my App factory and using `Activator.CreateInstance` instead. 
Any ideas?

**Comment by kekekeks** [comment, marked as answer]
Override `CreateAppBuilder` instead.

https://github.com/AvaloniaUI/Avalonia/blob/890df998a9273ec25e682585346b1e76678b4c1f/src/iOS/Avalonia.iOS/AvaloniaAppDelegate.cs#L36-L37

Not sure if it was intended to be used this way though.

@maxkatz6

**Reply by maxkatz6** [reply]
@kekekeks I don't remember/know why CustomizeAppBuilder initially was made this way, but it's only usable to customize app builder as @AArnott mentions.
`CreateAppBuilder` was added only recently and backported to 11.1.

**Reply by AArnott** [reply]
Thanks. 

If `CreateAppBuilder` isn't available before 11.1, that means there is no stable release that lets me provide the App factory, correct? Amazing that the need for it must have only come up recently.

**Reply by maxkatz6** [reply]
I haven't seen many usages of custom app constructor, and we don't use that in our samples. So, I guess that's it.

**Reply by AArnott** [reply]
That's good to know. Maybe I need to reevaluate the way I initialize my App.
